src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+                Task readDataAvailableTask = _responseDataAvailable.Task;
+
+                await _connection.SendFramesAsync(_requestBuffer.ReadMemory).ConfigureAwait(false);
+                _requestBuffer.Consume(_requestBuffer.ReadBytes.Length);
+
+                // Wait for response headers to be read.
+                await readDataAvailableTask.ConfigureAwait(false);
+
+                // TODO: Handle empty response using EmptyReadStream
+
+                responseContent.SetStream(new Http2ResponseStream(this));
+
+                return _response;
+            }
+
+            private static readonly byte[] StatusHeaderName = Encoding.ASCII.GetBytes(":status");
+
+            // Copied from HttpConnection
+            // TODO: Consolidate this logic?
+            private static bool IsDigit(byte c) => (uint)(c - '0') <= '9' - '0';
+
+            public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+            {
+                if (name.SequenceEqual(StatusHeaderName))
+                {
+                    if (value.Length != 3)
+                        throw new Exception("Invalid status code");
+
+                    // Copied from HttpConnection
+                    byte status1 = value[0], status2 = value[1], status3 = value[2];
+                    if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                    {
+                        throw new HttpRequestException(SR.net_http_invalid_response);
+                    }
+
+                    _response.SetStatusCodeWithoutValidation((HttpStatusCode)(100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0')));
+                }
+                else
+                {
+                    if (!HeaderDescriptor.TryGet(name, out HeaderDescriptor descriptor))
+                    {
+                        // Invalid header name
+                        throw new HttpRequestException(SR.net_http_invalid_response);
+                    }
+
+                    string headerValue = descriptor.GetHeaderValue(value);
+
+                    // Note we ignore the return value from TryAddWithoutValidation; 
+                    // if the header can't be added, we silently drop it.
+                    if (descriptor.HeaderType == HttpHeaderType.Content)
+                    {
+                        _response.Content.Headers.TryAddWithoutValidation(descriptor, headerValue);
+                    }
+                    else
+                    {
+                        _response.Headers.TryAddWithoutValidation(descriptor, headerValue);
+                    }
+                }
+            }
+
+            public void OnResponseHeadersComplete(bool endStream)
+            {
+                if (endStream)
+                {
+                    _responseComplete = true;
+                }
+
+                TaskCompletionSource<bool> readDataAvailable = _responseDataAvailable;
+                _responseDataAvailable = null;
```

Comment:
This doesn't need synchronization around the getting/setting of _responseDataAvailable?

Reply:
No. The initial _responseDataAvailable is constructed before we send the request. So it's guaranteed to exist at this point. The waiting code (SendAsync) won't look at these again until after the Task is signaled.